### PR TITLE
Sk/workflow take2

### PR DESCRIPTION
**Pull request by snopoke**
http://manage.dimagi.com/default.asp?239100
This was a doozy! Best reviewed by commit :fishing_pole_and_fish: 

@orangejenny @emord 
buddy @biyeun 
cc @phillipm 

Summary: In order to fix the bug we needed more information when generating the app workflow in the suite file e.g. what case type a session datum referenced. Adding that extra data broke some existing logic and required re-working most of the case list form workflow logic. Thankfully there is now more code reuse and a better 'algorithm' for generating the workflow logic.

### CODE REVIEW

File: corehq/apps/app_manager/suite_xml/post_process/workflow.py [hunk 1]
```diff
@@ -319,81 +367,64 @@ def case_list_forms_frames(self, form):
...
-            frame_case_created.add_child(root_module_command)
-            frame_case_not_created.add_child(root_module_command)
-
-            source_form_dm = add_datums_for_target(root_module, source_form_dm, allow_missing=True)
-
-        frame_case_created.add_child(CommandId(target_command))
-        frame_case_not_created.add_child(CommandId(target_command))
-        add_datums_for_target(target_module, source_form_dm)
-
+        stack_frames = CaseListFormStackFrames(case_created=frame_case_created, case_not_created=frame_case_not_created)
```
Comment (dimagimon) [inline corehq/apps/app_manager/suite_xml/post_process/workflow.py]: E501 line too long (120 > 115 characters)

File: corehq/apps/app_manager/tests/test_case_list_form.py [hunk 2]
```diff
@@ -156,6 +156,22 @@ def test_case_list_form_parent_child_basic(self):
...
+        factory.form_opens_case(register_person_form, case_type='stub', is_subcase=True)
+
+        person_module, update_person_form = factory.new_basic_module(
+            'update_person',
+            'person',
+            case_list_form=register_person_form
+        )
+
+        factory.form_requires_case(update_person_form)
+        self.assertXmlPartialEqual(self.get_xml('case_list_form_reg_form_creates_child_case'), factory.app.create_suite(), './entry[1]')
```
Comment (dimagimon) [inline corehq/apps/app_manager/tests/test_case_list_form.py]: E501 line too long (136 > 115 characters)

File: corehq/apps/app_manager/tests/test_form_workflow.py [hunk 3]
```diff
@@ -259,11 +259,8 @@ def test_link_to_form_in_parent_module(self):
         self.assertXmlPartialEqual(self.get_xml('form_link_child_modules'), factory.app.create_suite(), "./entry[3]")
 
     def test_form_links_submodule(self):
-        """
-        Test that when linking between two forms in a submodule we match up the session variables between the source
-        and target form correctly
-        :return:
-        """
+        # Test that when linking between two forms in a submodule we match up the session variables between the source
```
Comment (dimagimon) [inline corehq/apps/app_manager/tests/test_form_workflow.py]: E501 line too long (118 > 115 characters)

Comment (emord) [inline corehq/apps/app_manager/tests/test_form_workflow.py]: Curious why you decided that a comment was better than a docstring?

**Review by emord — APPROVED** [review]

**Comment by snopoke** [comment]
:white_check_mark:

**Comment by snopoke** [comment]
Because the test runner displays docstrings instead of the test function
name which I don't find useful

**Pull request merged**